### PR TITLE
Plugin - Shell Script

### PR DESCRIPTION
**Pull request by all3kcis**
Shell plugin, with alias overload.
more choice of execution programs
See https://github.com/davidmckenzie/pagermon/issues/130#issuecomment-443366926

### CODE REVIEW

**Comment by all3kcis** [comment]
Ok for me

File: server/plugins/Shell.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+var fs = require('fs');
+
+
+function run(trigger, scope, data, config, callback) {
+
+    var available_language = {
+      'sh':'sh',
+      'php':'/usr/bin/php',
+      'python':'/usr/bin/python',
+      'python3':'/usr/bin/python3',
```
[the remark this answers is not in the document]
Comment (all3kcis) [reply server/plugins/Shell.js]: Okay, I'll change that in that way.

File: server/plugins/Shell.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        var file_name = data.pluginconf.Shell.overrideAlias+'.'+used_ext;
+      }
+
+      
+
+      
+   
+      if(fs.existsSync(file_path+file_name)){
+          console.log('Exec shell command for selected alias')
+          dir = exec(used_program + ' '+file_path+file_name, function(err, stdout, stderr) {
```
[the remark this answers is not in the document]
Comment (DanrwAU) [reply server/plugins/Shell.js]: You might not be, but is there any use case where someone else might find that useful? 
I think the message is an important part and would be an good to be able to pass to an shell script for many cool things

[the remark this answers is not in the document]
Comment (mitch2k) [reply server/plugins/Shell.js]: Would be great to be able to have the message itself passed to the shell script.  I want to create a wall display/LED banner that displays the message and/or pass it to another system that reads out the message.

**Comment by all3kcis** [comment]
The data is now passed to the script.
first : address
two : message
three : json data object (all elements, alias, color, source, ...)
I couldn't test the execution of the powershell script.

**Comment by marshyonline** [comment]
@DanrwAU is the PowerShell Master IIRC :P 
Changes look good to be besides what has been suggested by @davidmckenzie

**Comment by DanrwAU** [comment]
> @DanrwAU is the PowerShell Master IIRC :P
> Changes look good to be besides what has been suggested by @davidmckenzie

I won't have time to test till Tuesday night but i'll try and put it through it's paces before then if i can 😄

**Comment by all3kcis** [comment]
> One other thing though @all3kcis - could you convert the logging statements to use appropriate log levels? E.g. `logger.main.debug()` and `logger.main.error()`

Ok NP

**Comment by all3kcis** [comment]
PowerShell works now :)
Capture  

2.ps1 :  
```
echo "Hello World"


$capcode = $args[0]
$message = $args[1]
$something = $args[2]

echo $capcode
echo $message
echo $something
```

**Comment by DanrwAU** [comment]
I've tested this one out and it seems to work as designed on powershell now. 

I still think the documentation/labels inside the GUI need to be clarified, possibly just a link to the wiki and explainining everything there?

@pagermon/collaborators happy for this to merge now?

**Comment by DanrwAU** [comment]
Is there a reason this needs a seperate gitignore file? could this be added to the one in the root?

**Comment by all3kcis** [comment]
> Is there a reason this needs a seperate gitignore file? could this be added to the one in the root?
  
This is to keep the spirit of the plugin system.
It is used to not tracking .sh/.ps1 files

**Comment by DanrwAU** [comment]
Good point :+1:

**Comment by all3kcis** [comment]
I can not edit wiki but you can add following text, edit it as you like : 

```markdown
The Shell plugin allows you to run your own scripts (bash / powershell) for the desired aliases.
- Enable the plugin in "Settings" 
- Create or edit the desired alias
 - Enable shell plugin for this alias and create file named {ID}.sh (Linux) or {ID}.ps1 (Windows)
**Files must be stored into "pagermon/server/plugins/Shell/"**
**Ex :** "5.sh", 5 represents the alias ID
 - To use the same script file as another alias, enter the alias ID of the file to be used in the "Override Alias" field
 > The script receives 3 arguments, <address (Integer)> <message (Text)> <data (JSON)>
 > Data contains many elements like, Id, timestamp, alias, agency, icon, color,  pluginconf...
```

**Comment by s3m1s0n1c** [comment]
> ```gfm
> The Shell plugin allows you to run your own scripts (bash / powershell) for the desired aliases.
> - Enable the plugin in "Settings" 
> - Create or edit the desired alias
>  - Enable shell plugin for this alias and create file named {ID}.sh (Linux) or {ID}.ps1 (Windows)
> **Files must be stored into "pagermon/server/plugins/Shell/"**
> **Ex :** "5.sh", 5 represents the alias ID
>  - To use the same script file as another alias, enter the alias ID of the file to be used in the "Override Alias" field
>  > The script receives 3 arguments, <address (Integer)> <message (Text)> <data (JSON)>
>  > Data contains many elements like, Id, timestamp, alias, agency, icon, color,  pluginconf...
> ```

I'll add this to wiki under plugins

**Comment by all3kcis** [comment]
I'm ready to merge :)

**Comment by DanrwAU** [comment]
Hope you don't mind @all3kcis i modified the desccriptions to include a link to the wiki and be a little more simple for simple people like me. 
Happy for this to merge now 👍

**Comment by davidmckenzie** [comment]
Forgot changelog.md? :D

**Comment by DanrwAU** [comment]
> Forgot changelog.md? :D

Fixed it up. 😁

**Comment by all3kcis** [comment]
Thanks to you :)

**Pull request merged**